components/prism-sas.js

```diff
@@ -10,13 +10,132 @@ Prism.languages.sas = {
 			'punctuation': /;/
 		}
 	},
+	'procSQL': {
+		pattern: /(^|\n)(?:proc\s+sql(?:\s+[\w|=]+)?);([\s\S]+(?:\r?\n|\r)(?=((?:proc\s\w+|quit|run|data(?!\=));))((?:proc\s\w+|quit|run|data(?!\=));)|(?:\r?\n|\r)[\s\S]+;)/im,
+		inside: {
+			'fullStep': {
+				pattern: /(^|\n)(?:proc\s+sql(?:\s+[\w|=]+)?);/i,
+				inside: {
+					'step': {
+						pattern: /(?:proc\s\w+|quit|run|data(?!\=))\b/i,
+						alias: 'keyword'
+					},
+					'argument': {
+						pattern: /\w+(?==)/,
+						alias: 'keyword'
+					},
+					'operator': /=/,
+					'punctuation': /;/,
+					'number': /\b(?:[\da-f]+x(?!\w+)|\d+(?:\.\d+)?(?:e[+-]?\d+)?)/,
+					'string': {
+						pattern: /(["'])(?:\1\1|(?!\1)[\s\S])*\1/,
+						greedy: true
+					}
+				}
+			},
+			'step': {
+				pattern: /(?:proc\s\w+|quit|run|data(?!\=))\b/i,
+				alias: 'keyword'
+			},
+			rest: Prism.languages.sql
+		}
+	},
+	'macrodeclaration': {
+		pattern: /((^|\n)(%macro)((\s)+([\w\s\S]+?))+);/im,
+		inside: {
+			keyword: /%macro/im,
+		}
+	},
+	'macroend': {
+		pattern: /((^|\n)(%mend)((\s)+([\w\s\S]+?))+);/im,
```

Comment:
Multiple issues:

1. `(^|\n)` because of the `m` flag `^` is equivalent to `(?<!.)`, or roughly `(?<=[\r\n])`. So I don't quite understand your intention here.
2. `((\s)+([\w\s\S]+?))+` with a few braces less `(\s+[\w\s\S]+?)+`: Because of this part your pattern will show exponential backtracking. 
  You can see for yourself using tools like [regexr](https://regexr.com/) using the pattern `(\s+[\w\s\S]+?)+;`. The red `ERROR` means that your pattern took longer than 250ms to execute.
  Don't worry, this is easy to fix because `(\s+[\w\s\S]+?)+` is equivalent to `[\s\S]+`, but I guess you really meant `[\s\S]+?`.
3. `[\w\s\S]` === `[\s\S]`

```suggestion
		pattern: /^%mend[\s\S]+?;/im,
```

Reply:
Yep. that looks right. Thanks.